### ISSUE TITLE
Confusing ifDefined extension error when author wants default property fallback behavior

### ISSUE THREAD

**Issue by AndrewJakubowicz** [runem/lit-analyzer#223]
### Context

The extension doesn't allow correct usage of `ifDefined` (I think). See minimal example and [playground link](https://lit.dev/playground/#project=W3sibmFtZSI6InNpbXBsZS1ncmVldGluZy50cyIsImNvbnRlbnQiOiIvKipcbiAqIEBsaWNlbnNlXG4gKiBDb3B5cmlnaHQgMjAyMSBHb29nbGUgTExDXG4gKiBTUERYLUxpY2Vuc2UtSWRlbnRpZmllcjogQlNELTMtQ2xhdXNlXG4gKi9cblxuIGltcG9ydCB7aHRtbCwgY3NzLCBMaXRFbGVtZW50LCBub3RoaW5nfSBmcm9tICdsaXQnO1xuIGltcG9ydCB7Y3VzdG9tRWxlbWVudCwgcHJvcGVydHl9IGZyb20gJ2xpdC9kZWNvcmF0b3JzLmpzJztcbiBpbXBvcnQge2lmRGVmaW5lZH0gZnJvbSAnbGl0L2RpcmVjdGl2ZXMvaWYtZGVmaW5lZC5qcyc7XG4gXG4gQGN1c3RvbUVsZW1lbnQoJ3NpbXBsZS1ncmVldGluZycpXG4gZXhwb3J0IGNsYXNzIFNpbXBsZUdyZWV0aW5nIGV4dGVuZHMgTGl0RWxlbWVudCB7XG4gICBzdGF0aWMgc3R5bGVzID0gY3NzYHAgeyBjb2xvcjogYmx1ZSB9YDtcbiBcbiAgIEBwcm9wZXJ0eSgpXG4gICBuYW1lPzogc3RyaW5nID0gJ0RlZmF1bHRWYWx1ZSc7XG4gXG4gICByZW5kZXIoKSB7XG4gICAgIHJldHVybiBodG1sYDxwPkhlbGxvLCAke3RoaXMubmFtZX0hPC9wPmA7XG4gICB9XG4gfVxuIFxuIEBjdXN0b21FbGVtZW50KCdpZi1kZWZpbmVkLXRlc3QnKVxuIGNsYXNzIFBhcmVudFRlc3QgZXh0ZW5kcyBMaXRFbGVtZW50IHtcbiAgIHJlbmRlcigpIHtcbiAgICAgY29uc3QgZGVmaW5lZE5hbWUgPSBcIkZvb1wiO1xuICAgICBjb25zdCB1bmRlZmluZWROYW1lID0gdW5kZWZpbmVkO1xuICAgICByZXR1cm4gaHRtbGBcbiAgICAgICAgIDxwPid1bmRlZmluZWQnIHRlc3Q6PC9wPlxuICAgICAgICAgPHNpbXBsZS1ncmVldGluZyAubmFtZT0ke3VuZGVmaW5lZE5hbWV9Pjwvc2ltcGxlLWdyZWV0aW5nPlxuIFxuICAgICAgICAgPCEtLSBUeXBlICdTeW1ib2wodW5kZWZpbmVkKScgaXMgbm90IGFzc2lnbmFibGUgdG8gJ3N0cmluZyB8IHVuZGVmaW5lZCdsaXQtcGx1Z2luKG5vLWluY29tcGF0aWJsZS10eXBlLWJpbmRpbmcpKDIzMDQpIC0tPlxuICAgICAgICAgPHA-J25vdGhpbmcnIHRlc3Q6PC9wPlxuICAgICAgICAgPHNpbXBsZS1ncmVldGluZyAubmFtZT0ke25vdGhpbmd9Pjwvc2ltcGxlLWdyZWV0aW5nPlxuIFxuICAgICAgICAgPHA-RGVmaW5lZCB0ZXN0OjwvcD5cbiAgICAgICAgIDxzaW1wbGUtZ3JlZXRpbmcgLm5hbWU9JHtkZWZpbmVkTmFtZX0-PC9zaW1wbGUtZ3JlZXRpbmc-XG4gXG4gICAgICAgICA8IS0tIFRoZSAnaWZEZWZpbmVkJyBkaXJlY3RpdmUgaGFzIG5vIGVmZmVjdCBoZXJlLiBsaXQtcGx1Z2luKG5vLWludmFsaWQtZGlyZWN0aXZlLWJpbmRpbmcpKDIzMDgpIC0tPlxuICAgICAgICAgPHA-aWZEZWZpbmVkKHVuZGVmaW5lZCkgdGVzdDo8L3A-XG4gICAgICAgICA8c2ltcGxlLWdyZWV0aW5nIC5uYW1lPSR7aWZEZWZpbmVkKHVuZGVmaW5lZE5hbWUpfT48L3NpbXBsZS1ncmVldGluZz5cbiAgICAgYDtcbiAgIH1cbiB9In0seyJuYW1lIjoiaW5kZXguaHRtbCIsImNvbnRlbnQiOiI8IURPQ1RZUEUgaHRtbD5cbjxoZWFkPlxuICA8c2NyaXB0IHR5cGU9XCJtb2R1bGVcIiBzcmM9XCIuL3NpbXBsZS1ncmVldGluZy5qc1wiPjwvc2NyaXB0PlxuPC9oZWFkPlxuPGJvZHk-XG4gIDxpZi1kZWZpbmVkLXRlc3Q-PC9pZi1kZWZpbmVkLXRlc3Q-XG48L2JvZHk-XG4ifSx7Im5hbWUiOiJwYWNrYWdlLmpzb24iLCJjb250ZW50Ijoie1xuICBcImRlcGVuZGVuY2llc1wiOiB7XG4gICAgXCJsaXRcIjogXCJeMi4wLjBcIixcbiAgICBcIkBsaXQvcmVhY3RpdmUtZWxlbWVudFwiOiBcIl4xLjAuMFwiLFxuICAgIFwibGl0LWVsZW1lbnRcIjogXCJeMy4wLjBcIixcbiAgICBcImxpdC1odG1sXCI6IFwiXjIuMC4wXCJcbiAgfVxufSIsImhpZGRlbiI6dHJ1ZX1d).

Explanation of the repro:

We have a custom element `simple-greeting` with optional `name` property with fallback `"DefaultValue"`. The example tests 4 cases:

1. Passing `undefined` directly to `.name`. This is valid but prevents the default value from rendering. I think this is expected.
2. Passing `nothing` to `.name`. Plugin raises error: `Type 'Symbol(undefined)' is not assignable to 'string | undefined'`. But we get correct behavior of the default value.
3. Passing a string directly. This works.
4. Using `ifDefined` raises plugin error:  `The 'ifDefined' directive has no effect here.lit-plugin(no-invalid-directive-binding)(2308)`. However correct rendering behavior is observed.

Here is a code sample that can be copy pasted into VS code to reproduce the plugin error:

```
 import {html, css, LitElement, nothing} from 'lit';
 import {customElement, property} from 'lit/decorators.js';
 import {ifDefined} from 'lit/directives/if-defined.js';
 
 @customElement('simple-greeting')
 export class SimpleGreeting extends LitElement {
 
   @property()
   name?: string = 'DefaultValue';
 
   render() {
     return html`<p>Hello, ${this.name}!</p>`;
   }
 }
 
 @customElement('if-defined-test')
 class ParentTest extends LitElement {
   render() {
     const definedName = "Foo";
     const undefinedName = undefined;
     return html`
         <p>'undefined' test:</p>
         <simple-greeting .name=${undefinedName}></simple-greeting>
 
         <!-- Type 'Symbol(undefined)' is not assignable to 'string | undefined'lit-plugin(no-incompatible-type-binding)(2304) -->
         <p>'nothing' test:</p>
         <simple-greeting .name=${nothing}></simple-greeting>
 
         <p>Defined test:</p>
         <simple-greeting .name=${definedName}></simple-greeting>
 
         <!-- The 'ifDefined' directive has no effect here. lit-plugin(no-invalid-directive-binding)(2308) -->
         <p>ifDefined(undefined) test:</p>
         <simple-greeting .name=${ifDefined(undefinedName)}></simple-greeting>
     `;
   }
 }
```

### Impact

This was raised in a chat as a point of confusion. Maybe there's a workaround that could be added to the error message.

**Comment by 43081j** [comment]
because ifDefined is intended to be used with attribute bindings IIRC.

meaning:

```html
<my-element

  .name=${ifDefined(val)} <- wrong

  name=${ifDefined(val)} <- right
>
```

**Comment by arthurevans** [comment]
In Lit 2, `ifDefined` is a simple function that returns the value or `nothing`.  @justinfagnani merged some code to support Lit 2 directives, but I don't think the plugin has been updated yet--see https://github.com/runem/lit-analyzer/issues/219#issuecomment-1018714267

**Comment by AndrewJakubowicz** [comment]
Capturing some offline conversation here. The behavior here is more nuanced because the case I have described is only on initial render. Once a reactive update has overwritten the default, we don't get the default back on future updates that pass in `nothing`. So except for a property binding on initial render, this would indeed be a no-op.

Example: https://lit.dev/playground/?mods=gists#gist=11f037fa531cb794b37b6c2a7fba917c

Closing this issue since I think the analyzer is correct.
